### PR TITLE
#3353 Implement MSV3 availability query and purchase order

### PR DESCRIPTION
**Pull request by metas-ts**
https://github.com/metasfresh/metasfresh/issues/3353

### CODE REVIEW

File: src/main/java/de/metas/ui/web/handlingunits/HUEditorRow.java [hunk 1]
```diff
@@ -288,28 +287,14 @@ public HUEditorRowAttributes getAttributes() throws EntityNotFoundException
...
-				.stream()
-				.map(includedRow -> streamRecursive(includedRow))
-				.reduce(Stream.of(row), Stream::concat);
-	}
-
 	public Optional<HUEditorRow> getIncludedRowById(final DocumentId rowId)
 	{
 		return streamRecursive()
 				.filter(row -> rowId.equals(row.getId()))
+				.map(row -> (HUEditorRow)row)
```
Comment (teosarca) [inline src/main/java/de/metas/ui/web/handlingunits/HUEditorRow.java]: pls use HUEditorRow.cast

File: src/main/java/de/metas/ui/web/handlingunits/HUEditorViewBuffer.java [hunk 2]
```diff
@@ -93,9 +93,10 @@ default boolean matchesAnyRowRecursive(final HUEditorRowFilter filter)
 		final HUEditorRow topLevelRow = getById(topLevelRowId.toDocumentId());
 		return topLevelRow
 				.streamRecursive()
+				.map(row -> (HUEditorRow)row)
```
Comment (teosarca) [inline src/main/java/de/metas/ui/web/handlingunits/HUEditorViewBuffer.java]: pls use HUEditorRow.cast

File: src/main/java/de/metas/ui/web/handlingunits/HUEditorViewBuffer_FullyCached.java [hunk 3]
```diff
@@ -341,7 +341,10 @@ private final boolean isRowIdIncluded(final Set<HUEditorRowId> parentRowIds, fin
 
 		public Stream<HUEditorRow> streamRecursive()
 		{
-			return rows.stream().flatMap(row -> row.streamRecursive());
+			return rows.stream()
+					.map(row -> row)
+					.flatMap(row -> row.streamRecursive())
```
Comment (teosarca) [inline src/main/java/de/metas/ui/web/handlingunits/HUEditorViewBuffer_FullyCached.java]: u can use row::streamRecursive

File: src/main/java/de/metas/ui/web/handlingunits/HUEditorViewBuffer_FullyCached.java [hunk 4]
```diff
@@ -341,7 +341,10 @@ private final boolean isRowIdIncluded(final Set<HUEditorRowId> parentRowIds, fin
...
+					.map(row -> (HUEditorRow)row);
```
Comment (teosarca) [inline src/main/java/de/metas/ui/web/handlingunits/HUEditorViewBuffer_FullyCached.java]: pls use HUEditorRow.cast

File: src/main/java/de/metas/ui/web/handlingunits/HUEditorViewBuffer_HighVolume.java [hunk 5]
```diff
@@ -207,7 +207,9 @@ private ViewEvaluationCtx getViewEvaluationCtx()
 		}
 
 		return streamPage(0, STREAM_ALL_MAX_SIZE_ALLOWED, filter, defaultSelection.getOrderBys())
-				.flatMap(row -> row.streamRecursive());
+				.map(row -> row)
+				.flatMap(row -> row.streamRecursive())
+				.map(row -> (HUEditorRow)row);
```
Comment (teosarca) [inline src/main/java/de/metas/ui/web/handlingunits/HUEditorViewBuffer_HighVolume.java]: pls use HUEditorRow.cast

File: src/main/java/de/metas/ui/web/handlingunits/HUEditorRow.java (hunk 1 above)
Comment (metas-ts) [reply src/main/java/de/metas/ui/web/handlingunits/HUEditorRow.java]: thx, done

File: src/main/java/de/metas/ui/web/handlingunits/HUEditorViewBuffer.java (hunk 2 above)
Comment (metas-ts) [reply src/main/java/de/metas/ui/web/handlingunits/HUEditorViewBuffer.java]: thx, done

File: src/main/java/de/metas/ui/web/handlingunits/HUEditorViewBuffer_FullyCached.java (hunk 4 above)
Comment (metas-ts) [reply src/main/java/de/metas/ui/web/handlingunits/HUEditorViewBuffer_FullyCached.java]: hahaaa i used HUEditorRow::cast instead !

File: src/main/java/de/metas/ui/web/handlingunits/HUEditorViewBuffer_HighVolume.java (hunk 5 above)
Comment (metas-ts) [reply src/main/java/de/metas/ui/web/handlingunits/HUEditorViewBuffer_HighVolume.java]: done

**Pull request merged**